### ISSUE TITLE
Is the number of train samples from tree nodes correct?

### ISSUE THREAD

**Issue by tlapusan** [tensorflow/decision-forests#105]
Hi, 

I'm trying to integrate tf-df into dtreeviz library for tree visualisations.

While working for this I observed that the number of samples from default visualisations (plot_model) is not the same with my own calculations.

This is the tree plot:
<img width="741" alt="Screenshot 2022-05-10 at 15 37 55" src="https://user-images.githubusercontent.com/12815158/167629989-bbd233f6-523f-44c3-b595-5cfc5e80e3ca.png">

And I checked the number of samples for child nodes from root node.
<img width="407" alt="Screenshot 2022-05-10 at 15 33 36" src="https://user-images.githubusercontent.com/12815158/167630130-9513cd48-7100-4bcc-bb9f-78269990e00e.png">

You can see that the number samples is different. For example the left child node should contains 623 samples, instead of 641. 
Is this because of some bagging sampling strategy ?

Here is how the tree structure will look in dtreeviz... but we still have to fix some little issues until it will be ready for release.
<img width="785" alt="Screenshot 2022-05-10 at 15 40 58" src="https://user-images.githubusercontent.com/12815158/167630557-925d0a40-d895-4b94-ba2d-22f0a073a72a.png">

**Comment by Cheril311** [comment]
@tlapusan can you please tell what hyperparameters you used and what tree-based model you used?

**Comment by tlapusan** [comment]
Indeed, I should mentioned them in the beginning :)
I'm using Mac OS (Montery) if this helps also.

<img width="767" alt="Screenshot 2022-05-11 at 10 27 34" src="https://user-images.githubusercontent.com/12815158/167793156-ccf8802a-f010-446b-8ff7-f6c4de385c50.png">

**Comment by achoum** [comment]
Hi @tlapusan,

The number of examples printed in the plot is the number of examples that reached a particular node during training. This is effectively the number of training examples *for the tree*. The training examples for *a tree* can be different from the training examples *for a model*.

For example, by default, when training a Random Forest, the training examples of each tree are [bagged](https://fr.wikipedia.org/wiki/Bootstrap_aggregating) from the original training dataset. While the number of examples is the same, the value of the examples are different.

For Gradient Boosted Trees and CART, by default, unless you provide a validation dataset in `fit` or if you disable early stopping, some of the training examples are effectively used for validation (and then virtually removed from the training dataset).


In your example, you can try replacing your model definition by:

```python
model = tfdf.keras.CartModel(validation_ratio=0.0)
```

Ps: Awesome for the dtreeviz integration. Please keep us in touch!

**Comment by tlapusan** [comment]
Hi @achoum,

thanks for your answer, it makes totally sense. I had also the assumption that the bagging could be the cause. Thanks for confirming this.

Sure, we will keep you up to date with dtreeviz integration :)

**Comment by tlapusan** [comment]
@achoum we are almost done with the integration in dtreeviz for tf.keras.RandomForestModel. You can take a look here https://github.com/tlapusan/dtreeviz/blob/support_for_TF-DF_trees_%23181/notebooks/dtreeviz_tensorflow_visualisations.ipynb

We have a little issue with the categorical features and the fact that TF-DF is converting them internally to integers. You can have more details about it from #111 .

Do you know how to find out which are the integer values associated to a categorical feature by TF-DF ? I need this information to make dtreeviz methods to work internally.
